openedx/core/djangoapps/discussions/docs/decisions/0001-discussion-configuration-api.rst

```diff
@@ -0,0 +1,65 @@
+Discussion provider configuration
+=================================
+
+Status
+------
+
+Proposal
+
+
+Context
+-------
+
+During the development of discussions plugins, a need arose for a system of
+configuration that would allow admins to configure connections to different
+discussion tools, and have course authors select an established configuration
+without requiring admins to share any secrets with course authors.
+
+For example, a discussion tool provider might require admins to specify OAuth2
+credentials that will be needed to integrate their tool with the platform. This
+needs to be configured by an admin, and used by a course, but ideally course
+authors should not have access to this information.
+
+It would also be useful to have multiple sets of configuration for different
+discussion tools. A course creator can then select one of the available
+configurations based on their requirements.
+
+For instance an organisation might be using a particular Discourse instance
+across a number of their courses. So an admin will create a new configuration
+object that specifies all the settings the Discourse plugin needs to operate,
+such as the site location, the auth credentials etc. When setting up a new
+course, a course author will then be able to see this as one of the possible
+options when configuring Discourse as the discussion tool.
+
+Additionally, we might want to prevent certain configurations to be available
+to all courses on a site. So it would be useful to limit configuration options
+to a specific site, or Organization.
+
+This ADR proposes a new configuration system that allows admins to create
+pre-populated configurations for specific discussions providers, which can then
+be used and optionally overridden by course authors/admins.
+
+Decision
+--------
+
+We can create two models, `DiscussionProviderConfig` and
+`LearningContextDiscussionConfig` that together provide the functionality
+described above.
+
+The `DiscussionProviderConfig` model stores a configuration name (this will be
+visible to course authors configuring a discussion tool), a provider (this will
+be the id of the plugin for which the configuration is defined), and a config
+JSON field (this will store the actual config). A configuration can be also be
+restricted to only be visible/available for a specific site, or org by setting
+`restrict_to_site` or `restrict_to_org`. Only one of these can be set at a time
```

Comment:
Thinking ahead a bit - what do you think the effort would be to add other types of restrictions in the future, given the current model?  i.e., `restrict_to_program`, for instance.  I'm not at all saying we need to do this now, just thinking about future extensibility as we have shared discussions forums for various types of groups.

Reply:
I think that will depend on how such future content is structured. My understanding is more of these relationships will be formalised in the database going forwards, so adding more restrictions should be straightforward. I only see complexity if we want to allow some combination of restrictions, but not others. 

For the specific case of programs since that relationship is in an external DB it definitely will complicate these queries. If the end goal is to have the same discussion provider and discussion instance across all courses in the program, then that would be accomplished by linking a `LearningContextDiscussionConfig` to the entire program via a program key. 

If the goal instead is that the config should only be available for use for courses inside a program, then perhaps it might make sense to cache the program key course key relation in edx-platform as well.